docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,444 @@
+
+public:
+    SecureStore(KVStore *underlying_kv, KVStore *rbp_kv);
+    virtual ~SecureStore();
+    	 
+    // Initialization and formatting
+    virtual int init();
+    virtual int deinit();
+    virtual int reset();
+
+    // Core API
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, size_t offset = 0);
+    virtual int get_info(const char *key, info_t *info);
+    virtual int remove(const char *key);
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+    virtual int set_finalize(set_handle_t handle);
+ 
+    // Key iterator
+    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size, size_t *actual_key_size);
+    virtual int iterator_close(iterator_t it);
+    
+private:
+    Mutex _mutex;
+    KVStore *_underlying_kv;
+    KVStore *_rbp_kv;
+    void *_entropy;
+    uint8_t *_scratch_buf;
+    int _unique_key_id_counter;  
+}
+```
+
+### Important data structures
+
+```C++
+// Record header
+typedef struct {
+	uint16_t metadata_size;
+	uint16_t reserved;
+	uint32_t data_size;
+	uint32_t create_flags;
+	uint32_t unique_key_id;
+	uint8_t  nonce[8];
+} record_metadata_t;
+
+// incremental set handle
+typedef struct {
+    record_metadata_t metadata;
+    bd_size_t offset;
+    void *encrypt_handle;
+    void *auth_handle;
+    KVStore::set_handle_t underlying_handle;
+} inc_set_handle_t;
+
+// iterator handle
+typedef struct {
+    KVStore::iterator_t underlying_it;
+} key_iterator_handle_t;
+```
+
+
+### Initialization and reset
+
+**init function**
+
+Header:  
+`virtual int init();`
+
+Pseudo code:  
+- if `_is_initialized` return OK
+- Take `_mutex`
+- Initialize `_entropy` with TLS entropy APIs
+- Using `DeviceKey` APIs, get the device key
+- Allocate `_scratch_buf` as a 32 byte array
+- Iterate over all keys is `_rbp_kv` and set `_unique_key_id_counter` to the maximal key + 1   
+- Set `_is_initialized` to true
+- Release `_mutex`     
+  
+**deinit function**
+
+Header:  
+`virtual int deinit();`
+
+Pseudo code:  
+- if not `_is_initialized` return OK
+- Take `_mutex` 
+- Deinitialize `_entropy`
+- Deallocate `_encrypt_key`, `_auth_key`, `_scratch_buf`
+- Release `_mutex`   
+
+**reset function**
+
+Header:  
+`virtual int reset();`
+
+Pseudo code:  
+- Take `_mutex` 
+- Call `_underlying_kv` `reset` API  
+- Call `_rbp_kv` `reset` API  
+- Set `_unique_key_id_counter` to 1
+- Release `_mutex`   
+
+### Core APIs
+
+**set function**
+
+Header:  
+`virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);`
+
+Pseudo code:  
+- Call `set_start` with all fields and a local `set_handle_t` variable
```

Comment:
RBP logic is handled in set_start and set_finalize. Set function just calls set_start, set_add_data and set_finalize in order to prevent duplication of code.

Reply:
ok. I missed that you are calling set_start in SecureStore and not in tdbstore

Reply:
I do it in both cases.